@xen-orchestra/backups/RemoteAdapter.js

```diff
@@ -460,22 +472,84 @@ class RemoteAdapter {
       await asyncMap(files, async file => {
         try {
           const metadata = await this.readVmBackupMetadata(file)
-          if (predicate === undefined || predicate(metadata)) {
-            // inject an id usable by importVmBackupNg()
-            metadata.id = metadata._filename
-
-            backups.push(metadata)
-          }
+          // inject an id usable by importVmBackupNg()
+          metadata.id = metadata._filename
+          backups[file] = metadata
         } catch (error) {
-          warn(`listVmBackups ${file}`, { error })
+          warn(`createCacheListVmBackups ${file}`, { error })
         }
       })
+      return backups
     } catch (error) {
       let code
       if (error == null || ((code = error.code) !== 'ENOENT' && code !== 'ENOTDIR')) {
         throw error
       }
     }
+  }
+
+  // use _ to mark this method as private
+  // since we decorate it with syncrhonized.withKey in constructor
+  // and # function are not writeable
+
+  async _createCacheListVmBackups(vmUuid) {
+    const path = `${BACKUP_DIR}/${vmUuid}/cache.json.gz`
+    try {
+      const backups = await this.#getCachabledDataListVmBackups(vmUuid)
+      if (backups === undefined) {
+        return
+      }
+      const text = JSON.stringify(backups)
+      const zipped = await pFromCallback(zlib.gzip, text)
+
+      // cache file may be updated by multiple XO on the same remote
+      // we try to overwrite the cache file, and if it results
+      // to a broken cache file , it will be invalidated on next read
+      await this.handler.writeFile(path, zipped, { flags: 'w' })
+
+      return backups
+    } catch (error) {
+      let code
+      if (error == null || ((code = error.code) !== 'ENOENT' && code !== 'ENOTDIR')) {
+        throw error
+      }
+    }
+  }
+
+  // read the list of backup of a Vm from cache
+  // if cache is broken => delete it and regenerate it
+  // if cache is missing => regenerate it
+
+  async #readCacheListVmBackups(vmUuid) {
```

Comment:
Your code is nice but I think I still prefer the cache handling to be isolated in a single function (which could be named `readOrCacheVmBackups`).